[yunchao] iP

tanyunchao opened the pull request:
# Martin-the-best-chatbot
> "Coding is hard, but naming variables is harder" [source](chat.openai.com)

## Types of events Martin can handle:
- Todos
- Events
- Deadlines 

Other features:
1. Mark and unmark task 
2. Saves to a **dedicated** file 
3. Able to delete tasks 😄 `delete task 3`

## Heres how to download
- [ ] Download java
- [ ] Fork and download

Run using 
```
java Martin
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

tanyunchao committed:
Rename and refactor

tanyunchao committed:
Renamed bot

tanyunchao committed:
added echo function

tanyunchao committed:
feat: listing function

tanyunchao committed:
refactor: added task class, update list type

tanyunchao committed:
added comments

tanyunchao committed:
fix: changed tick icon

tanyunchao committed:
feat: mark and unmark tasks

tanyunchao committed:
refactor: new task types inheriting from task

tanyunchao committed:
feat: use enums

tanyunchao committed:
fix: added UNMARK

tanyunchao committed:
feat: break tasks and use enum

tanyunchao committed:
updated test scripts

tanyunchao committed:
feat: added error throwing

tanyunchao committed:
refactor: removed unused imports

tanyunchao committed:
feat: added delete function

tanyunchao committed:
feat: added txt file system

tanyunchao committed:
fix: access level for class fields

tanyunchao committed:
revert: made todolist static again

tanyunchao committed:
Implemented startUpSequence

tanyunchao committed:
add file string

tanyunchao committed:
implement writing to file

tanyunchao committed:
chore: simple imports and comment clean up

tanyunchao committed:
Merge branch 'branch-Level-7'

tanyunchao committed:
Fix deadline string formatting

tanyunchao committed:
Add date fornatting for deadline

tanyunchao committed:
Merge branch 'branch-Level-8'

tanyunchao committed:
Creat new dummy classes

tanyunchao committed:
Refactor Martin to start becoming an instance

tanyunchao committed:
Initialise Storage class and associated methods

tanyunchao committed:
Add startupSequence method

tanyunchao committed:
Initialise TaskList class

tanyunchao committed:
Add task marking functions

tanyunchao committed:
Add functionality into parser

tanyunchao committed:
Initialise command class

tanyunchao committed:
Add greeting and farewell messages to Ui class

tanyunchao committed:
Add getRemainingWords method to Parser class

tanyunchao committed:
Refactor how Martin fundamentally works

teojunda commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + this.due.format(DateTimeFormatter.ofPattern(DATE_TIME_FORMAT)) + ")";
```
Perhaps it is more suitable to make the line break before the operator?

teojunda commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+
+    public String[] deadlineParse(String input) {
```
Perhaps it is better to phrase the method as a verb, like `parseDeadline`?

teojunda commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,49 @@
+        return new String[]{deadlineDescription, deadlineArray[1]};
+    }
+
+    public String[] eventParse(String input) {
```
Perhaps it is better to phrase the method as a verb, like `parseEvent`?

tanyunchao replied:
thanks for raising this up @teojunda , I think its definitely a valid point that i missed

teojunda commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,49 @@
+    public ChatbotKeyword parse(String input) {
+        String[] inputArray = input.split(" ");
+        String command = inputArray[0].toLowerCase();
+        switch (command) {
+            case "list":
+                return ChatbotKeyword.LIST;
+            case "done":
+                return ChatbotKeyword.MARK;
+            case "delete":
+                return ChatbotKeyword.DELETE;
+            case "todo":
+                return ChatbotKeyword.TODO;
+            case "deadline":
+                return ChatbotKeyword.DEADLINE;
+            case "event":
+                return ChatbotKeyword.EVENT;
+            case "bye":
+                return ChatbotKeyword.BYE;
+            default:
+                throw new IllegalArgumentException("I'm sorry, but I don't know what that means :-(");
```
I really like the use of switch case statements and enumerations!

teojunda commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,22 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task{
```
Perhaps you can follow add a space before '{'?

teojunda reviewed (commented):
Looks good 👍 
Overall, I like how readable your code was! However, I noticed a few standard coding violations. Also, it adding some JavaDocs to explain the classes and methods would greatly improve the code clarity.

tanyunchao committed:
Add martin.txt to gitignore

tanyunchao committed:
Refactor code to use objects and append to file

tanyunchao committed:
Refactor Martin class initialization and startup sequence

tanyunchao committed:
Update Parser.java with new command options

tanyunchao committed:
Fix writing tasks to file in Storage class

tanyunchao committed:
Merge remote-tracking branch 'origin/add-gradle-support'

tanyunchao committed:
Update .gitignore to include martin.txt

tanyunchao committed:
Repackage as martin package

tanyunchao committed:
Add ParserTest class with test cases for getRemainingWords method

tanyunchao committed:
Add TaskListTest class for unit testing

tanyunchao committed:
Add package declaration and import statements

tanyunchao committed:
Remove unused variable in Martin class

tanyunchao committed:
Add Martin class documentation

tanyunchao committed:
Add parser documentation

tanyunchao committed:
Add documentation to storage

tanyunchao committed:
Add greeting and goodbye messages to Ui class

tanyunchao committed:
Refactor Command class to include Javadoc comments

tanyunchao committed:
Refactor method names in Command and Parser classes

tanyunchao committed:
Add FIND keyword to ChatbotKeyword enum

tanyunchao committed:
Add FIND command to search for tasks

tanyunchao committed:
Add find() method to TaskList class

tanyunchao committed:
Add printFoundTasks method to Ui class

tanyunchao committed:
Fix file writing in Storage class

tanyunchao committed:
Add support for 'find' command in Parser class

tanyunchao committed:
Merge branch 'branch-A-JavaDoc'

tanyunchao committed:
Merge branch 'branch-A-CodingStandard'

tanyunchao committed:
Merge branch 'branch-Level-9'

chiaryan commented on src/main/java/Martin/Parser.java:
```diff
@@ -0,0 +1,85 @@
+        String[] inputArray = input.split(" ");
+        String command = inputArray[0].toLowerCase();
+        switch (command) {
+            case "list":
```
Should the indenting for these `case` blocks be changed to align with the Java [coding standard?](https://se-education.org/guides/conventions/java/intermediate.html#layout)

chiaryan commented on src/test/java/Martin/ParserTest.java:
```diff
@@ -0,0 +1,37 @@
+
+public class ParserTest {
+    @Test
+    public void testGetRemainingWords() {
```
Should this be renamed to follow the naming convention for Junit tests?

chiaryan commented on src/test/java/Martin/ParserTest.java:
```diff
@@ -0,0 +1,37 @@
+        assertEquals(expectedOutput1, parser.getRemainingWords(input1));
+
+        // Test case 2: input with multiple words
+        String input2 = "Hello world, how are you?";
```
Perhaps you can consider wrapping your tests 2-5 in separate test functions, so that you can run them independently from each other?

chiaryan commented on src/test/java/Martin/TaskListTest.java:
```diff
@@ -0,0 +1,33 @@
+        taskList.add(new Todo("Task 3"));
+
+        // Redirect the standard output to a ByteArrayOutputStream
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
```
Oo, didn't know you could redirect sysout like this, very nice way to test the print function.

chiaryan commented on src/main/java/Martin/Command.java:
```diff
@@ -0,0 +1,93 @@
+                break;
+            case MARK:
+                int doneIndex = Integer.parseInt(remainingWords);
+                taskList.markAsDone(doneIndex);
```
For methods that access the elements of the Tasklist, Is there any exception handler for the case where it tries to access an out-of bounds index?

tanyunchao replied:
At this current point no. But definitely a possible extension of the project, since index out of bounds definitely happens. Thanks for the heads up

chiaryan reviewed (commented):
Good job, couldn't find any issues w.r.t. coding standard. See my comments for possible improvements.

tanyunchao committed:
Add JavaFX dependencies and implement JavaFX Application

tanyunchao committed:
Add Launcher class to workaround classpath issues

tanyunchao committed:
Add new images

tanyunchao committed:
Add DialogBox class to handle user and Duke dialogues

tanyunchao committed:
Add chat UI components and functionality

tanyunchao committed:
Remove unused method getDialogLabel()

tanyunchao committed:
Add imports and update DialogBox class

tanyunchao committed:
Add DialogBox and MainWindow FXML files

tanyunchao committed:
Update Launcher class to use Main instead of Martin

tanyunchao committed:
Add Main class for GUI implementation

tanyunchao committed:
Add MainWindow class with controller functionality

tanyunchao committed:
Refactor Martin class and remove JavaFX dependencies

tanyunchao committed:
Update Martin variable name in MainWindow class

tanyunchao committed:
Delete martin.txt file

tanyunchao committed:
Refactor Command class constructor and handleCommand method

tanyunchao committed:
Refactor getResponse method to call run method

tanyunchao committed:
Refactored TaskList class methods to return String instead of void

tanyunchao committed:
Refactor Ui class methods to return strings instead of printing messages

tanyunchao committed:
Merge branch 'branch-Level-10'

tanyunchao committed:
Refactor initialization of objects in Martin class

tanyunchao committed:
Update mainClass and archiveBaseName in build.gradle

tanyunchao committed:
Enable assertions in run task

tanyunchao committed:
Remove unused import statement and main method in Martin.java

tanyunchao committed:
Add null assertion for tasks

Makes sure that the tasklist is in fact initiated after the start up and when Martin is constructed.

tanyunchao committed:
Add javadocs to all TaskList methods

Javadocs is important as it generates documentation for future user or maintainers.

This also allows IDEs to show the explaination of methods on hover.

tanyunchao committed:
Fix code style

tanyunchao committed:
Assert size of todolist on initiation

tanyunchao committed:
Merge pull request #1 from tanyunchao/branch-A-Assertions

Assertions and long commits

tanyunchao committed:
Add documentation for Deadline class

tanyunchao committed:
Simplify code structure for strings

tanyunchao committed:
Simplify string structure

tanyunchao committed:
Refactor Martin class to improve readability and functionality

tanyunchao committed:
Remove redundant run() method in Martin class

tanyunchao committed:
Rename getRemainingWords to removeCommandWord

tanyunchao committed:
Generate docs

tanyunchao committed:
Clean up code

tanyunchao committed:
Add docs

tanyunchao committed:
Refactor to improve code quality

tanyunchao committed:
Add docs

tanyunchao committed:
Add docs to Ui methods

tanyunchao committed:
Merge pull request #2 from tanyunchao/branch-A-CodeQuality

Code Quality improvements

tanyunchao committed:
Feat: implement help function

tanyunchao committed:
Merge pull request #3 from tanyunchao/branch-BCD-Extension

Feat: implement help function

tanyunchao committed:
Add directory creation if it doesn't exist in Storage.java

tanyunchao committed:
Add greeting functionality to ChatbotKeyword and Command classes

tanyunchao committed:
Refactor printList method to handle empty list case

tanyunchao committed:
Add exception handling for invalid date format

tanyunchao committed:
Update Event class to use hyphen in the date range

tanyunchao committed:
Refactor ParserTest.java to remove redundant test cases

tanyunchao committed:
Add test cases for Parser class

tanyunchao committed:
Delete TaskListTest.java

tanyunchao committed:
Add Ui.png file

tanyunchao committed:
Update user guide with daily tasks examples and improve help command

tanyunchao committed:
update package naming

tanyunchao committed:
Add EventTest class with test cases

tanyunchao committed:
Add TodoTest class with test cases

tanyunchao committed:
Add DeadlineTest class with test cases

tanyunchao committed:
Handle exception in MainWindow.java
